tests/cuda/test_sync.cu

```diff
@@ -0,0 +1,97 @@
+/**
+ * TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2021 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#include <gtest/gtest.h>
+
+#include "../../device/cuda/src/utils/sync.cuh"
+
+__global__ void testWarpIndexedBallotSyncBasicKernel(uint32_t *vts,
+                                                     uint32_t *vis) {
+    auto [vt, vi] =
+        traccc::cuda::warp_indexed_ballot_sync(threadIdx.x % 2 == 0);
+
+    vts[threadIdx.x] = vt;
+    vis[threadIdx.x] = vi;
+}
+
+__global__ void testWarpIndexedBallotSyncWithExitKernel(uint32_t *vts,
+                                                        uint32_t *vis) {
+    if (threadIdx.x < 16) {
+        return;
+    }
+
+    auto [vt, vi] =
+        traccc::cuda::warp_indexed_ballot_sync(threadIdx.x % 2 == 0);
```

Comment:
Great question! When you issue a synchronization instruction in CUDA, you can pass it a 32-bit mask, and threads will wait for all _non-exited_ threads for which the corresponding bit in the mask is true to pass through that barrier. In other words, a barrier will not wait for a thread that is exited, even though it might be included in the mask. This, no doubt, was implemented this way to reduce the risk of running into dead-lock situations. In this particular program, the masks `0b11111111111111111111111111111111` and `0b00000000000000001111111111111111` are equivalent.

Masks can still be useful if you want to synchronize only part of a warp!

Reply:
Ah, I see, thank you!